Issue #10473 - Revert jetty.sh pgrep, and update started checks

joakime opened the pull request:
Revert the jetty.sh pgrep usage.
Add distribution test for jetty.conf.
Update "started" function in jetty.sh.
Ensure "state" module executes before setuid or deploy kicks in.
Ensure state file is initialized early.

joakime committed:
Revert pgrep usage in jetty.sh

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Adding test for jetty.conf

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Correcting renamed xml file

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Improved started check

+ Improved `started` function code
  - adding comment explaining steps
  - adding named parameters
+ Improved ARGS check for "jetty.state="
  option, to know when to check the
  state file.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Make sure state.mod is before any deploy steps to ensure jetty-state file is created early.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

danwashusen commented:
Taking the `jetty.sh` and `state.mod` files from this patch, applying it to a 10.0.16 extract and creating a Jetty base with the following seems to have made things worse. In that I'm unable to start Jetty at all:
```sh
java -jar {{ jetty.home }}/start.jar --create-startd
            --add-module="server,state,deploy,http,ssl,https,proxy-protocol,websocket,http-forwarded,jsp,jstl,ext,resources,webapp"
```

The Jetty output logs (trimmed):
```
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 Reading /etc/default/jetty..
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JAVA                  =  /usr/bin/java
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JAVA_OPTIONS          =  -javaagent:/usr/local/jetty/newrelic/newrelic.jar ...
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_HOME            =  /usr/local/jetty
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_BASE            =  /var/lib/jetty
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 START_D               =  /var/lib/jetty/start.d
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 START_INI             =  /var/lib/jetty/start.ini
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_START           =  /usr/local/jetty/start.jar
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_CONF            =  /usr/local/jetty/etc/jetty.conf
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_ARGS            =  jetty.state=/var/lib/jetty/state jetty.pid=/run/jetty/jetty.pid --module=pid,state
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_RUN             =  /run/jetty
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_PID             =  /run/jetty/jetty.pid
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_START_LOG       =  /var/log/jetty/start.log
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_STATE           =  /var/lib/jetty/state
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_START_TIMEOUT   =  180
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_SYS_PROPS       =
2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 RUN_ARGS              =  -Djava.io.tmpdir=/tmp -Djetty.home=/usr/local/jetty -Djetty.base=/var/lib/jetty -javaagent:/usr/local/jetty/newrelic/newrelic.jar ... --class-path /var/lib/jetty/resources:/usr/local/jetty/lib/logging/slf4j-api-2.0.5.jar:/usr/local/jetty/lib/logging/jetty-slf4j-impl-10.0.16.jar:/usr/local/jetty/lib/jetty-servlet-api-4.0.6.jar:/usr/local/jetty/lib/jetty-http-10.0.16.jar:/usr/local/jetty/lib/jetty-server-10.0.16.jar:/usr/local/jetty/lib/jetty-xml-10.0.16.jar:/usr/local/jetty/lib/jetty-util-10.0.16.jar:/usr/local/jetty/lib/jetty-io-10.0.16.jar:/usr/local/jetty/lib/jetty-jndi-10.0.16.jar:/usr/local/jetty/lib/jetty-security-10.0.16.jar:/usr/local/jetty/lib/jetty-servlet-10.0.16.jar:/usr/local/jetty/lib/jetty-webapp-10.0.16.jar:/usr/local/jetty/lib/jetty-plus-10.0.16.jar:/usr/local/jetty/lib/jakarta.transaction-api-1.3.3.jar:/usr/local/jetty/lib/jetty-annotations-10.0.16.jar:/usr/local/jetty/lib/annotations/asm-9.5.jar:/usr/local/jetty/lib/annotations/asm-analysis-9.5.jar:/usr/local/jetty/lib/annotations/asm-commons-9.5.jar:/usr/local/jetty/lib/annotations/asm-tree-9.5.jar:/usr/local/jetty/lib/annotations/jakarta.annotation-api-1.3.5.jar:/usr/local/jetty/lib/apache-jsp/org.eclipse.jdt.ecj-3.26.0.jar:/usr/local/jetty/lib/apache-jsp/org.eclipse.jetty.apache-jsp-10.0.16.jar:/usr/local/jetty/lib/apache-jsp/org.mortbay.jasper.apache-el-9.0.52.jar:/usr/local/jetty/lib/apache-jsp/org.mortbay.jasper.apache-jsp-9.0.52.jar:/usr/local/jetty/lib/apache-jstl/org.apache.taglibs.taglibs-standard-impl-1.2.5.jar:/usr/local/jetty/lib/apache-jstl/org.apache.taglibs.taglibs-standard-spec-1.2.5.jar:/usr/local/jetty/lib/jetty-client-10.0.16.jar:/usr/local/jetty/lib/jetty-alpn-client-10.0.16.jar:/usr/local/jetty/lib/jetty-deploy-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-core-common-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-core-client-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-core-server-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-servlet-10.0.16.jar:/usr/local/jetty/lib/websocket/jetty-javax-websocket-api-1.1.2.jar:/usr/local/jetty/lib/websocket/websocket-javax-client-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-javax-common-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-javax-server-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-jetty-api-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-jetty-common-10.0.16.jar:/usr/local/jetty/lib/websocket/websocket-jetty-server-10.0.16.jar org.eclipse.jetty.xml.XmlConfiguration ... org.eclipse.jetty.util.log.class=org.eclipse.jetty.util.log.StdErrLog runtime.feature.alpn=true slf4j.version=2.0.5 spring.profiles.active=region-apau,region-apsg,region-euie,region-uswest,scheduler,redis-rate-limiter sun.net.inetaddr.ttl=10 /usr/local/jetty/etc/jetty-bytebufferpool.xml /usr/local/jetty/etc/jetty-pid.xml /usr/local/jetty/etc/jetty-threadpool.xml /usr/local/jetty/etc/jetty.xml /usr/local/jetty/etc/jetty-webapp.xml /usr/local/jetty/etc/jetty-state.xml /usr/local/jetty/etc/jetty-deploy.xml /usr/local/jetty/etc/jetty-http.xml /usr/local/jetty/etc/jetty-http-forwarded.xml /usr/local/jetty/etc/jetty-ssl.xml /usr/local/jetty/etc/jetty-ssl-context.xml /usr/local/jetty/etc/jetty-https.xml /usr/local/jetty/etc/jetty-proxy-protocol.xml
2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
2023-09-20T02:10:08.435000+00:00 i-0b51d4668e8996463 FAILED Wed Sep 20 02:10:03 UTC 2023
```

Confusingly now it doesn't even seem to attempt to deploy our app...

joakime committed:
Early cleanup / creation of State file

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Improved `started` function

+ Don't attempt to read from State File
  if it doesn't exist
+ Don't attempt to read from PID File
  if it doesn't exist
+ DEBUG in state file logic
+ DEBUG in pid file logic
+ proper startsWith logic for state detection

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented:
> ```
> 2023-09-20T02:10:00.060000+00:00 i-0b51d4668e8996463 JETTY_START_TIMEOUT   =  180
> 2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
> 2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
> 2023-09-20T02:10:04.068000+00:00 i-0b51d4668e8996463 tail: cannot open '/var/lib/jetty/state' for reading: No such file or directory
> 2023-09-20T02:10:08.435000+00:00 i-0b51d4668e8996463 FAILED Wed Sep 20 02:10:03 UTC 2023
> ```

You had a FAILED response in 8.4 seconds on a 180 second timeout.
That would indicate that the startup failed, and the '/var/lib/jetty/state' reported it as such.
The next step for you would be to look at your logs in `/var/lib/jetty/logs/*` and see what kind of failure it was.

I'm deploying a new 10.0.17-SNAPSHOT that will have additional DEBUG options present for the state file startup ATM, I'll update here when it is available.

joakime commented:
The latest 10.0.17-SNAPSHOT that contains the changes from this PR can be found at ...

https://oss.sonatype.org/content/repositories/jetty-snapshots/org/eclipse/jetty/jetty-home/10.0.17-SNAPSHOT/jetty-home-10.0.17-20230920.122955-20.tar.gz

Please enable DEBUG for both the jetty.sh and the State file behaviors when testing this jetty-home.

If using `jetty-slf4j-impl`, add the following line to `resources/jetty-logging.properties`
```
org.eclipse.jetty.server.StateLifeCycleListener.LEVEL=DEBUG
```

If using another slf4j implementation, set the named logger `org.eclipse.jetty.server.StateLifeCycleListener` to DEBUG level.

When running `jetty.sh` add the `-d` option to enable DEBUG.
Eg:
``` shell
/path/to-jetty-home/bin/jetty.sh -d start
```

danwashusen commented:
@joakime confirming that when using the linked snapshot everything behaves as expected.

joakime committed:
Better state debugging and pid kill

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

danwashusen commented:
@joakime One thing to note is that we are seeing the following message from systemd when using the linked snapshot:
```
2023-09-21T16:22:09.000000+00:00 i-XYZ Sep 21 16:22:09 foobar systemd[1]: jetty.service: Supervising process 4431 which is not our child. We'll most likely not notice when it exits.
```

janbartel commented:
@joakime this branch doesn't work when the `demo` module is enabled.
I did:
* mkdir base; cd base
* java -jar ../jetty-home/start.jar --add-modules=demo
* ../jetty-home/bin/jetty.sh start

Output is:

```
ERROR : Unknown module='demo'. List available with --list-modules

Usage: java -jar $JETTY_HOME/start.jar [options] [properties] [configs]
       java -jar $JETTY_HOME/start.jar --help  # for more information
Starting Jetty: Usage: java [options] <mainclass> [args...]
           (to execute a class)
   or  java [options] -jar <jarfile> [args...]
           (to execute a jar file)
   or  java [options] -m <module>[/<mainclass>] [args...]
       java [options] --module <module>[/<mainclass>] [args...]
           (to execute the main class in a module)
   or  java [options] <sourcefile> [args]
           (to execute a single source-file program)

 Arguments following the main class, source file, -jar <jarfile>,
 -m or --module <module>/<mainclass> are passed as the arguments to
 main class.

 where options include:

    -cp <class search path of directories and zip/jar files>
    -classpath <class search path of directories and zip/jar files>
    --class-path <class search path of directories and zip/jar files>
                  A : separated list of directories, JAR archives,
                  and ZIP archives to search for class files.
    -p <module path>
    --module-path <module path>...
                  A : separated list of directories, each directory
                  is a directory of modules.
    --upgrade-module-path <module path>...
                  A : separated list of directories, each directory
                  is a directory of modules that replace upgradeable
                  modules in the runtime image
    --add-modules <module name>[,<module name>...]
                  root modules to resolve in addition to the initial module.
                  <module name> can also be ALL-DEFAULT, ALL-SYSTEM,
                  ALL-MODULE-PATH.
    --enable-native-access <module name>[,<module name>...]
                  modules that are permitted to perform restricted native operations.
                  <module name> can also be ALL-UNNAMED.
    --list-modules
                  list observable modules and exit
    -d <module name>
    --describe-module <module name>
                  describe a module and exit
    --dry-run     create VM and load main class but do not execute main method.
                  The --dry-run option may be useful for validating the
                  command-line options such as the module system configuration.
    --validate-modules
                  validate all modules and exit
                  The --validate-modules option may be useful for finding
                  conflicts and other errors with modules on the module path.
    -D<name>=<value>
                  set a system property
    -verbose:[class|module|gc|jni]
                  enable verbose output for the given subsystem
    -version      print product version to the error stream and exit
    --version     print product version to the output stream and exit
    -showversion  print product version to the error stream and continue
    --show-version
                  print product version to the output stream and continue
    --show-module-resolution
                  show module resolution output during startup
    -? -h -help
                  print this help message to the error stream
    --help        print this help message to the output stream
    -X            print help on extra options to the error stream
    --help-extra  print help on extra options to the output stream
    -ea[:<packagename>...|:<classname>]
    -enableassertions[:<packagename>...|:<classname>]
                  enable assertions with specified granularity
    -da[:<packagename>...|:<classname>]
    -disableassertions[:<packagename>...|:<classname>]
                  disable assertions with specified granularity
    -esa | -enablesystemassertions
                  enable system assertions
    -dsa | -disablesystemassertions
                  disable system assertions
    -agentlib:<libname>[=<options>]
                  load native agent library <libname>, e.g. -agentlib:jdwp
                  see also -agentlib:jdwp=help
    -agentpath:<pathname>[=<options>]
                  load native agent library by full pathname
    -javaagent:<jarpath>[=<options>]
                  load Java programming language agent, see java.lang.instrument
    -splash:<imagepath>
                  show splash screen with specified image
                  HiDPI scaled images are automatically supported and used
                  if available. The unscaled image filename, e.g. image.ext,
                  should always be passed as the argument to the -splash option.
                  The most appropriate scaled image provided will be picked up
                  automatically.
                  See the SplashScreen API documentation for more information
    @argument files
                  one or more argument files containing options
    -disable-@files
                  prevent further argument file expansion
    --enable-preview
                  allow classes to depend on preview features of this release
To specify an argument for a long option, you can use --<name>=<value> or
--<name> <value>.
```

joakime commented:
@janbartel I cannot replicate what you are seeing on my own workstation, my laptop, or any of the 3 docker images I'm using.

Configuring jetty-base ...

``` 
[jetty-base]$ java -jar ../jetty-home-10.0.17-SNAPSHOT/start.jar -Dmaven.repo.uri=https://oss.sonatype.org/content/groups/jetty-with-staging/  --add-module=demo 
...(snip lots of output)...
INFO  : download https://repo1.maven.org/maven2/jakarta/transaction/jakarta.transaction-api/1.3.3/jakarta.transaction-api-1.3.3.jar to ${jetty.base}/lib/ext/jakarta.transaction-api-1.3.3.jar
INFO  : download https://repo1.maven.org/maven2/org/bouncycastle/bcpkix-jdk15on/1.70/bcpkix-jdk15on-1.70.jar to ${jetty.base}/lib/bouncycastle/bcpkix-jdk15on-1.70.jar
INFO  : download https://repo1.maven.org/maven2/org/bouncycastle/bcprov-jdk15on/1.70/bcprov-jdk15on-1.70.jar to ${jetty.base}/lib/bouncycastle/bcprov-jdk15on-1.70.jar
INFO  : download https://repo1.maven.org/maven2/org/bouncycastle/bcutil-jdk15to18/1.70/bcutil-jdk15to18-1.70.jar to ${jetty.base}/lib/bouncycastle/bcutil-jdk15on-1.70.jar
INFO  : Base directory was modified
```

Then running it ...

```
[jetty-base]$ ../jetty-home-10.0.17-SNAPSHOT/bin/jetty.sh start
Starting Jetty: 2023-09-22 20:40:24.056:DEBUG:oeju.PidFile:main: PID file: /var/www/test/jetty-base/jetty/jetty.pid
2023-09-22 20:40:24.216:WARN :oe.jetty:main: demo-realm is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:24.409:WARN :oejk.KeystoreGenerator:main: Generating Test Keystore: DO NOT USE IN PRODUCTION!
2023-09-22 20:40:24.859:INFO :oejs.Server:main: jetty-10.0.17-SNAPSHOT; built: 2023-09-22T14:08:23.336Z; git: c8e8ad5b3d22221229075c2decbad3995ed57988; jvm 17.0.8.1+1-Ubuntu-0ubuntu122.04
2023-09-22 20:40:24.913:INFO :oejus.SslContextFactory:main: x509=X509@51acdf2e(jetty-test-keystore,h=[localhost],a=[],w=[]) for Server@6a55299e[provider=null,keyStore=file:///var/www/test/jetty-base/etc/test-keystore.p12,trustStore=null]
2023-09-22 20:40:24.973:INFO :oejdp.ScanningAppProvider:main: Deployment monitor [file:///var/www/test/jetty-base/webapps/]
2023-09-22 20:40:25.304:INFO :oejss.DefaultSessionIdManager:main: Session workerName=node0
2023-09-22 20:40:25.323:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@6c25e6c4{root,/,file:///var/www/test/jetty-base/webapps/root/,AVAILABLE}{/var/www/test/jetty-base/webapps/root}
2023-09-22 20:40:25.374:WARN :oejshC.demo_jetty_xml:main: The test-jetty webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.442:INFO :oejshC.demo_jetty_xml:main: TestFilter#remote=false
2023-09-22 20:40:25.455:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@a776e{Test WebApp,/test,file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-jetty_war-_test-any-/webapp/,AVAILABLE}{/var/www/test/jetty-base/webapps/demo-jetty.war}
2023-09-22 20:40:25.456:INFO :oejsh.ContextHandler:main: Started o.e.j.s.h.MovedContextHandler@14b030a0{demo-moved-context.xml,/oldContextPath,null,AVAILABLE}
2023-09-22 20:40:25.485:WARN :oejshC.demo_jsp_war:main: The demo-jsp webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.519:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@41e350f1{Demo JSP Web Application,/demo-jsp,file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-jsp_war-_demo-jsp-any-/webapp/,AVAILABLE}{/var/www/test/jetty-base/webapps/demo-jsp.war}
2023-09-22 20:40:25.617:WARN :oejshC.demo_proxy_war:main: The test-proxy webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.771:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@422c3c7a{Transparent Proxy WebApp,/proxy,file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-proxy_war-_demo-proxy-any-/webapp/,AVAILABLE}{/var/www/test/jetty-base/webapps/demo-proxy.war}
2023-09-22 20:40:25.802:WARN :oejshC.demo_jaas_xml:main: The test-jaas webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.823:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@6b9ce1bf{JAAS Test,/test-jaas,file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-jaas_war-_test-jaas-any-/webapp/,AVAILABLE}{/var/www/test/jetty-base/webapps/demo-jaas.war}
2023-09-22 20:40:25.852:WARN :oejshC.demo_jndi_xml:main: The test-jndi webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.886:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@150ab4ed{Test JNDI WebApp,/test-jndi,file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-jndi_war-_test-jndi-any-/webapp/,AVAILABLE}{/var/www/test/jetty-base/webapps/demo-jndi.war}
2023-09-22 20:40:25.971:WARN :oejshC.demo_async_rest_war:main: The async-rest webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:25.992:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@6e1d8f9e{Async REST Webservice Example,/demo-async-rest,[file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-async-rest_war-_demo-async-rest-any-/webapp/, jar:file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-async-rest_war-_demo-async-rest-any-/webapp/WEB-INF/lib/demo-async-rest-jar-10.0.17-SNAPSHOT.jar!/META-INF/resources],AVAILABLE}{/var/www/test/jetty-base/webapps/demo-async-rest.war}
2023-09-22 20:40:26.031:WARN :oejshC.demo_spec_xml:main: The test-spec webapp is deployed. DO NOT USE IN PRODUCTION!
2023-09-22 20:40:26.031:INFO :oe.jetty:main: WEB-INF/lib/jetty-util.jar logging used!
2023-09-22 20:40:26.084:INFO :oejsh.ContextHandler:main: Started o.e.j.w.WebAppContext@4f071df8{Test Annotations WebApp,/test-spec,[file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-spec_war-_test-spec-any-/webapp/, jar:file:///var/www/test/jetty-base/work/jetty-0_0_0_0-8443-demo-spec_war-_test-spec-any-/webapp/WEB-INF/lib/demo-web-fragment-10.0.17-SNAPSHOT.jar!/META-INF/resources],AVAILABLE}{/var/www/test/jetty-base/webapps/demo-spec.war}
2023-09-22 20:40:26.092:INFO :oejs.AbstractConnector:main: Started ServerConnector@1639f93a{SSL, (ssl, alpn, h2, http/1.1)}{0.0.0.0:8443}
2023-09-22 20:40:26.123:INFO :oejs.AbstractConnector:main: Started ServerConnector@771d1ffb{HTTP/1.1, (http/1.1)}{0.0.0.0:8080}
2023-09-22 20:40:26.129:INFO :oejs.Server:main: Started Server@6581dc0a{STARTING}[10.0.17-SNAPSHOT,sto=5000] @2293ms
OK Fri Sep 22 20:40:27 UTC 2023
```

janbartel commented:
The only difference I can see is that I checked out the branch of this PR and built it locally. After that, here's the steps (all executed as the user 'janb', not 'root' BTW):

```
  181  cd jetty-home/target
  182  ls
  183  mkdir base; cd base
  184  java -jar ../jetty-home/start.jar --add-modules=demos
  185  java -jar ../jetty-home/start.jar --add-modules=demo
  186  java -jar ../jetty-home/start.jar 
  187  ../jetty-home/bin/jetty.sh start
```

I still get exactly the same error as before.

joakime commented:
> The only difference I can see is that I checked out the branch of this PR and built it locally. After that, here's the steps (all executed as the user 'janb', not 'root' BTW):
> 
> ```
>   181  cd jetty-home/target
>   182  ls
>   183  mkdir base; cd base
>   184  java -jar ../jetty-home/start.jar --add-modules=demos
>   185  java -jar ../jetty-home/start.jar --add-modules=demo
>   186  java -jar ../jetty-home/start.jar 
>   187  ../jetty-home/bin/jetty.sh start
> ```
> 
> I still get exactly the same error as before.

Still doesn't fail for me.
**BUT** I also setup the `/etc/default/jetty` properly before I run `jetty.sh`
As that's where the `JETTY_HOME` and `JETTY_BASE` directories come from.

Can you run `jetty.sh status` and paste them here?

janbartel commented:
@joakime here it is:

```
[205] ../jetty-home/bin/jetty.sh status
ERROR : Unknown module='demo'. List available with --list-modules

Usage: java -jar $JETTY_HOME/start.jar [options] [properties] [configs]
       java -jar $JETTY_HOME/start.jar --help  # for more information
Jetty NOT running

JAVA                  =  /home/janb/java/jvm/jdk19/bin/java
JAVA_OPTIONS          =  
JETTY_HOME            =  /home/janb/src/jetty-eclipse/jetty-9.4.x/jetty-distribution/target/distribution
JETTY_BASE            =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base
START_D               =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/start.d
START_INI             =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/start.ini
JETTY_START           =  /home/janb/src/jetty-eclipse/jetty-9.4.x/jetty-distribution/target/distribution/start.jar
JETTY_CONF            =  /home/janb/src/jetty-eclipse/jetty-9.4.x/jetty-distribution/target/distribution/etc/jetty.conf
JETTY_ARGS            =  jetty.state=/home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty.state jetty.pid=/home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty/jetty.pid jetty-started.xml
JETTY_RUN             =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty
JETTY_PID             =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty/jetty.pid
JETTY_START_LOG       =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty/jetty-start.log
JETTY_STATE           =  /home/janb/src/jetty-eclipse/jetty-10.0.x/jetty-home/target/base/jetty.state
JETTY_START_TIMEOUT   =  60
JETTY_SYS_PROPS       =  
RUN_ARGS              =  
```

Interesting that `JETTY_HOME` is not the parent of the `bin` directory where the `jetty.sh` script is. I thought the algorithm used to find `JETTY_HOME` looked there?

joakime commented:
> ```
> JETTY_HOME            =  /home/janb/src/jetty-eclipse/jetty-9.4.x/jetty-distribution/target/distribution
> JETTY_START           =  /home/janb/src/jetty-eclipse/jetty-9.4.x/jetty-distribution/target/distribution/start.jar
> ```

Yeah, that wont work.
My guess is that you likely have an existing `/etc/default/jetty` that points to the jetty-9.4.x directory.

This would also explain the error message you got ...

```
ERROR : Unknown module='demo'. List available with --list-modules

Usage: java -jar $JETTY_HOME/start.jar [options] [properties] [configs]
       java -jar $JETTY_HOME/start.jar --help  # for more information
```

As Jetty 9.4.x doesn't have a `demo` module.

joakime commented:
> Interesting that `JETTY_HOME` is not the parent of the `bin` directory where the `jetty.sh` script is. I thought the algorithm used to find `JETTY_HOME` looked there?

There is some effort done in the `jetty.sh` to figure out the `JETTY_HOME`, but only if it is not set. (I suspect that on your env the `JETTY_HOME` variable is set, either on the shell ENV or from the `/etc/default/jetty` file)

https://github.com/eclipse/jetty.project/blob/jetty-9.4.52.v20230823/jetty-home/src/main/resources/bin/jetty.sh#L220-L239

Then there is the common use case of copying (or symlinking) jetty.sh into a path like `/etc/init.d/jetty` or `/etc/rc.d/rc3/jetty`.
Which would break the linked attempt to find the JETTY_HOME when used that way.

janbartel commented:
@joakime bingo!  A 5yr old `/etc/default/jetty` file. I deleted that and things now work as expected.

sbordet commented on jetty-home/src/main/resources/bin/jetty.sh:
```diff
@@ -118,35 +118,86 @@ findDirectory()
+  # wait till timeout to see "STARTED" in state file, needs --module=state as argument
+  for ((T = 0; T < $(($STARTTIMEOUT / 4)); T++))
   do
     sleep 4
```
I'd sleep 1 and do a normal loop.

joakime replied:
We've had 4 here since before Jetty 6.

sbordet replied:
And elsewhere in this file we `sleep 1`, so I'd be consistent, and it's simpler.
I always hated the long pause before having a feedback when Jetty starts, so I'd prefer 1 second sleeps.

sbordet commented on jetty-home/src/main/resources/bin/jetty.sh:
```diff
@@ -118,35 +118,86 @@ findDirectory()
+      echo -n "-"
+    else
+      (( DEBUG )) && echo "Unable to read State File: $STATEFILE"
+      echo -n ":"
```
Is this echo in the right place?
Also, should not this branch return 1, as the file is not readable?

joakime replied:
We don't return here.
As the state file might be unreachable at first, but then show up in later iteration of the loop.

sbordet commented on jetty-home/src/main/resources/bin/jetty.sh:
```diff
@@ -497,22 +544,23 @@ case "$ACTION" in
           cd \"$JETTY_BASE\"
           echo ${RUN_ARGS[*]} start-log-file=\"$JETTY_START_LOG\" | xargs ${JAVA} > /dev/null &
-          disown $(pgrep -P $!)"
+          disown \$!"
```
There is an extra closing double quote?

joakime replied:
This double-quote closes the one opened on line 497

sbordet commented on jetty-home/src/main/resources/bin/jetty.sh:
```diff
@@ -118,35 +118,86 @@ findDirectory()
+        (( DEBUG )) && echo "PID=$PID is running, sending kill signal=KILL (TIMEOUT=$TIMEOUT)"
+        kill -KILL "$PID" 2>/dev/null
+      fi
+      sleep 1
```
print something to give feedback that we are waiting?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/StateLifeCycleListener.java:
```diff
@@ -37,53 +38,75 @@ public class StateLifeCycleListener implements LifeCycle.Listener
+    }
+
+    @Override
+    public String toString()
```
Move at end of class, as usual.

sbordet commented on tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DistributionTests.java:
```diff
@@ -109,6 +111,54 @@ public void testStartStop() throws Exception
+                assertEquals(HttpStatus.NOT_FOUND_404, response.getStatus());
+            }
+
+            await().atMost(Duration.ofSeconds(10)).until(() -> !Files.exists(pidfile));
```
I'd check also for the state file, whether it should still be there or not.

joakime replied:
The state file is only cleaned up by the jetty.sh (or equivalent jetty service file)

sbordet reviewed (changes requested):
Still think we should simplify the loop in `started()`.

joakime committed:
Changes from review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Changes from review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet commented on jetty-home/src/main/resources/bin/jetty.sh:
```diff
@@ -118,35 +118,88 @@ findDirectory()
+      echo -n "-"
+    else
+      (( DEBUG )) && echo "Unable to read State File: $STATEFILE"
+      echo -n ":"
```
A small nit here, when we wait we print ".", not sure why here we print ":"?

joakime replied:
In this loop, we print either ..
 * `-` which means we found the state file, but it doesn't contain one of the keywords that finishes the `started` loop.
 * `:` which means the state file wasn't found (yet)
 
We don't print `.` during the started loop.

Let me add a few more 1 char states to this loop, and a cleanup of the output while i'm at it.

joakime replied:
With the last commit the characters during the loop are ...

* `-` which means we found the state file, but it doesn't contain one of the keywords that finishes the `started` loop.
* `:` which means the state file wasn't found (yet)

And when the `started` check finishes, there are 4 terminations of the progress indicator.

* `.` - which means the started file was found, and STARTED was detected
* `!` - which means the started file was found, and STOPPED was detected
* `!!` - which means the started file was found, and FAILED was detected
* ` timeout` - which means the loop terminated due to timeout.

Now we can see something like ...

```
Starting :::---.
OK Fri Sep 22 20:40:27 UTC 2023
```

which tells me that it had 3 iterations of no started file, 3 iterations of a started file but without a state that finishes (could be something like STARTING), and finally a finished STARTED.

But if we see ...

```
Starting ::::::::---!!
FAILED Fri Sep 22 20:40:27 UTC 2023
```

We know that the started file existed and reported FAILED.

But if we see ...

```
Starting ::::::::::::::::::::::::::::::::::::::::::::::::::::::----- timeout
FAILED Fri Sep 22 20:40:27 UTC 2023
```

Then the started file never reached STARTED  / STOPPED / or FAILED
And the state file took a long time to even show up.
The startup failed due to timeout reached.

sbordet replied:
@joakime ok.

But `pidKill()` has `echo "."` to indicate wait, so to be consistent can you change that to be `:`, in a way similar to above?

joakime replied:
pidKill is actually much simpler too.

joakime committed:
More progress indicators of started() loop

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
More changes from review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>
